### PR TITLE
Follow-up from PR #1001 to try and express things more clearly

### PR DESCRIPTION
**Pull request by dcoutts**
In particular by moving the difference calculation into `addHeadersInFlight` and explaining things there.

Also added more commentary.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs [hunk 1]
```diff
@@ -488,12 +519,12 @@ newTMergeVar = TMergeVar <$> newEmptyTMVar
 -- | Merge the current value with the given one and store it, return the updated
 -- value.
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Stale comment still says `return the updated value`

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs [hunk 2]
```diff
@@ -249,6 +261,16 @@ newtype FetchRequest header =
...
+-- range request messages we send. We send one message per fragment. We never
+-- expect fetch requests to overlap (ie have blocks in common) but we do expect
+-- that frequently requests will \"touch\" so that two ranges could be merged
+-- into a single contigious range. This is useful because it lets us send fewer
+-- requests for bigger ranges.
+--
+-- So this semigroup instance does that style of merging. Otherwise the list
+-- of fragments is just appended.
+--
 instance HasHeader header => Semigroup (FetchRequest header) where
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Your new comment paragraphs are very clear. However, they also have a lot of overlap, all discussing the same concern (merging vs counting) in three different places. Maybe those places could be consolidated?

This instance, for example, could be folded into a more concrete function? Or maybe the queue (and so this instance) could explicitly track how many fragments it has merged?

These are mostly-idle suggestions, please take it or leave it :)

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs [hunk 3]
```diff
@@ -200,22 +200,34 @@ initialPeerFetchInFlight =
...
+      -- unacknowledged) fetch requests when the old and new turn out to be for
+      -- contigious ranges. This means that we end up sending fewer requests
+      -- but for bigger ranges. It also makes tracking the number of requests
+      -- in-flight a bit more subtle. We have to look at the /difference/ in
+      -- the number of fragments (corresponding to one request each) for the
+      -- existing (if any) and new request.
       peerFetchReqsInFlight   = peerFetchReqsInFlight inflight
-                              + fromIntegral newRequests,
+                              + fromIntegral (length fragments)
+                              - maybe 0 (length . fetchRequestFragments) existing,
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: `length . fetchRequests`, will produce an `Int`, but `peerFetchReqsInFlight :: Word`.

Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: I did say I'd not even compiled it :-)  (was having troubles with ghc versions). Thanks, will get those sorted and update the PR.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs (hunk 2 above)
Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Yeah, perhaps I can just refer to one of them, e.g. the semi-group instance.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs (hunk 1 above)
Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Ta.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: That's the only problem CI found.

**Comment by intricate** [comment]
@dcoutts: Is this clear to merge or are changes still needed?

**Comment by dcoutts** [comment]
This should address all the review feedback, and rebase on master.

**Comment by dcoutts** [comment]
Ok, I have apparently broken something. Have to go back and check.

**Comment by dcoutts** [comment]
So this proved to be even more subtle than I had appreciated. I hope it's clear in the code now!

It passes 10,000 QC iterations now locally.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs [hunk 4]
```diff
@@ -212,22 +212,37 @@ initialPeerFetchInFlight =
...
                    => (header -> SizeInBytes)
-                   -> FetchRequest header
-                   -> Int
-                   -- ^ the number of new requests; this might differ from the
-                   -- length of 'FetchRequest' by one, since when we write to
-                   -- @TFetchRequestVar@ we combine 'FetchRequest's using their
-                   -- semigroup instance, which might merge two chain fragments.
+                   -> Maybe (FetchRequest header) -- ^ Any existing request.
+                   -> FetchRequest header         -- ^ The new (merged) request.
+                   -> FetchRequest header         -- ^ The extra request.
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: The whole diff looks good to me. My only nitpick is here; I tripped over the use of the word "extra". Minor, just FYI.

We have (maybe) `old`, `new`, and `extra`, where `new = old <> extra` (or something close to that). But the word _extra_ initially sounds to me like "something else" that is not already in `new`. So maybe "latest"? "added"? "latestAdded"? Or so on.

Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Thanks. And that's the relationship, and is a nice concise way to put it. I'll add that.

**Comment by nfrisby** [comment]
bors cancel

**Comment by nfrisby** [comment]
Sorry -- you had asked for someone to merge it earlier, but I realized you might want to squash etc. Clearly, I 100% Approve this PR :)

**Pull request merged**